src/inmanta/parser/cache.py

```diff
@@ -0,0 +1,108 @@
+from inmanta.compiler.config import feature_compiler_cache
+from inmanta.parser.pickle import ASTPickler, ASTUnpickler
+from inmanta.util import get_compiler_version
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CacheManager:
+    def __init__(self) -> None:
+        self.hits = 0
+        self.misses = 0
+        self.failures = 0
+
+    def get_file_name(self, filename: str) -> Optional[str]:
+        """
+        Returns the name for the cached file, based on the name of the original source file
+
+        Also ensures the cache folder exists.
+
+        :param filename: the filename of the source file
+        :return: the filename of the cached file
+        """
+        base_folder = os.path.dirname(filename)
+        if not os.path.isdir(base_folder):
+            return None
+
+        filepart = os.path.basename(filename).rsplit(".", 1)[0]
+
+        cache_folder = os.path.join(base_folder, "__cfcache__")
+
+        os.makedirs(cache_folder, exist_ok=True)
+
+        filename = f"{filepart}.{get_compiler_version().replace('.','_')}.cfc"
+
+        return os.path.join(cache_folder, filename)
+
+    def un_cache(self, namespace: Namespace, filename: str) -> Optional[List[Statement]]:
+        if not feature_compiler_cache.get():
+            # cache not enabled
+            return None
+        try:
+            cache_filename = self.get_file_name(filename)
+            if not cache_filename:
+                self.misses += 1
+                return None
+            if not os.path.exists(cache_filename):
+                self.misses += 1
+                return None
+            if os.path.getmtime(filename) > os.path.getmtime(cache_filename):
+                self.misses += 1
+                return None
+            with open(cache_filename, "rb") as fh:
+                self.hits += 1
+                return ASTUnpickler(fh, namespace).load()
+        except Exception:
+            self.failures += 1
+            LOGGER.exception("Compile cache loading failure, ignoring cache entry for %s", filename)
+            return None
+
+    def cache(self, filename: str, statements: List[Statement]) -> None:
+
+        if not feature_compiler_cache.get():
+            # cache not enabled
+            return
+        try:
+            cache_filename = self.get_file_name(filename)
+            if not cache_filename:
+                # could not find file
+                return
+
+            with open(cache_filename, "wb") as fh:
+                ASTPickler(fh, protocol=4).dump(statements)
+        except Exception:
+            LOGGER.exception("Compile cache failure, failed to write for %s", filename)
+
+    def log_stats(self) -> None:
+        if not feature_compiler_cache.get():
+            # cache not enabled
+            return
+        LOGGER.info(
+            "Compiler cache observed %d hits and %d misses (%.d%%)",
+            self.hits,
+            self.misses,
+            (100 * self.hits) / (self.hits + self.misses),
```

Comment:
Shouldn't we also take into account the `failures` counter here?

Reply:
Should we reset the counters when `log_stats` is called?

Reply:
About failures: good question, to which I don't have an answer.

About resetting the counter: I'm opposed to that. 
If we also report these stats elsewhere (e.g. via the influxdb reporter), if every read resets, they will both get wrong data. 

In the current compiler, it is naturally reset, as it restarts every time. 

For other applications (language server, future compiler server), other aggregations or reset might be required.